model/labels/regexp.go

```diff
@@ -787,13 +789,50 @@ func (m *equalMultiStringMapMatcher) setMatches() []string {
 
 func (m *equalMultiStringMapMatcher) Matches(s string) bool {
 	if !m.caseSensitive {
-		s = strings.ToLower(s)
+		s = toNormalisedLower(s)
 	}
 
 	_, ok := m.values[s]
 	return ok
 }
 
+// toNormalisedLower normalise the input string using "Unicode Normalization Form D" and then convert
+// it to lower case.
+func toNormalisedLower(s string) string {
+	// Check if the string is all ASCII chars and convert any upper case character to lower case character.
+	isASCII := true
+	var (
+		b   strings.Builder
+		pos int
+	)
+	b.Grow(len(s))
+	for i := 0; i < len(s); i++ {
+		c := s[i]
+		if isASCII && c >= utf8.RuneSelf {
+			isASCII = false
+		}
+		if 'A' <= c && c <= 'Z' {
+			c += 'a' - 'A'
+			if pos < i {
+				b.WriteString(s[pos:i])
+			}
+			b.WriteByte(c)
+			pos = i + 1
+		}
+	}
+	if pos < len(s) {
+		b.WriteString(s[pos:])
+	}
+
+	// Optimize for ASCII-only strings. In this case we don't have to do any normalization.
+	if isASCII {
+		return b.String()
+	}
+
+	// Normalise and convert to lower.
+	return strings.Map(unicode.ToLower, norm.NFKD.String(b.String()))
```

Comment:
I think you can abort looping above once you find the first non-ascii char, and just do this here:

```suggestion
	return strings.Map(unicode.ToLower, norm.NFKD.String(s))
```

Reply:
Commited new changes

Reply:
But why build and allocate a new string if we aren't going to need it? IMO, we should optimize for ascii only, and then optimize for lowercase only.

If you allow me a suggestion, I'd write the method like this:

```go
// toNormalisedLower normalise the input string using "Unicode Normalization Form D" and then convert it to lower case.
// This method is optimized for ASCII-only strings.
func toNormalisedLower(s string) string {
	// Check if the string is all ASCII chars and convert any upper case character to lower case character.
	isASCII := true
	hasUpper := false
	for _, c := range s {
		hasUpper = hasUpper || ('A' <= c && c <= 'Z')
		if c >= utf8.RuneSelf {
			isASCII = false
			break
		}
	}
	if !isASCII {
		return strings.Map(unicode.ToLower, norm.NFKD.String(s))
	}
	if !hasUpper {
		return s
	}

	var (
		b   strings.Builder
		pos int
	)
	for i := 0; i < len(s); i++ {
		c := s[i]
		if 'A' <= c && c <= 'Z' {
			if pos < i {
				b.WriteString(s[pos:i])
			}
			c += 'a' - 'A'
			b.WriteByte(c)
			pos = i + 1
		}
	}
	if pos < len(s) {
		b.WriteString(s[pos:])
	}
	return b.String()
}
```

Reply:
I have a concern regarding the approach of iterating through a string twice to check for uppercase characters and then converting them to lowercase, especially for ASCII strings where this process might seem redundant. Instead, I propose a single loop solution to convert all uppercase characters to lowercase, regardless of whether the string is ASCII or not.
What are your thoughts on this alternative?

Reply:
I was assuming that unnecessarily allocating a new string (when everything is lowercase) is more expensive than looping through an ascii string twice, but of course that depends on the string length and we can only talk if we run the benchmarks. 

> I proposed a single loop solution to convert all uppercase characters to lowercase, regardless of whether the string is ASCII or not.

Did you mean _regardless of whether the string has uppercase or not_?

Reply:
We can have `toNormalisedLower` take a buffer (say 2KB) which we allocate on the stack of the caller, thus removing that allocation cost when the string is not already lowercase.

Reply:
> We can have `toNormalisedLower` take a buffer (say 2KB) which we allocate on the stack of the caller, thus removing that allocation cost when the string is not already lowercase.

Could you please clarify if you are referring to passing the input string as a buffer to `toNormalisedLower`?
Apologies, but I'm having difficulty understanding your explanation. Could you please clarify or rephrase it for me?

Reply:
IMO, we should commit to something working and not terrible at performance, and then we can nerd snipe each other with alternative implementations of this method, benchmarking specifically this method (instead of the entire matcher).

@Ranveer777 bhat @bboreham is suggesting is to define an array, say `var arr byte[2048]` and create a slice using that array `out := arr[:0]`, and build the string on that slice as we proceed, something like:
```go
func toNormalisedLower(s string) string {
	var (
		arr [2024]byte
		b   = bytes.NewBuffer(arr[:0])
		pos int
	)
	hasUpper := false
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= utf8.RuneSelf {
			return strings.Map(unicode.ToLower, norm.NFKD.String(s))
		}
		if 'A' <= c && c <= 'Z' {
			hasUpper = true
			if pos < i {
				b.WriteString(s[pos:i])
			}
			c += 'a' - 'A'
			b.WriteByte(c)
			pos = i + 1
		}
	}
	if !hasUpper {
		return s
	}
	if pos < len(s) {
		b.WriteString(s[pos:])
	}
	return b.String()
}
```

Reply:
> @Ranveer777 what @bboreham is suggesting is to define an array, say `var arr byte[2048]` and create a slice using that array `out := arr[:0]`, and build the string on that slice as we proceed, something like:

Thank you for clarifying. I initially perceived it as a suggestion for commited implementation.



> IMO, we should commit to something working and not terrible at performance, and then we can nerd snipe each other with alternative implementations of this method, benchmarking specifically this method (instead of the entire matcher).

👍